Assets/Fungus/Thirdparty/FungusLua/Scripts/ConversationManager.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using UnityEngine;
+
+namespace Fungus
+{
+	class ConversationManager : MonoBehaviour
+	{
+		protected static ConversationManager instance;
+
+		protected Stage stage;
+
+		protected Dictionary<string, Character> characters;
+
+		protected Character characterNotFound;
+
+		protected Character characterNone;
+
+		static public ConversationManager getInstance()
+		{
+			if (instance == null)
+			{
+				GameObject go = new GameObject("ConversationManager");
+				DontDestroyOnLoad(go);
+				instance = go.AddComponent<ConversationManager>();
+			}
+
+			return instance;
+		}
+
+		protected virtual void Awake ()
+		{
+			stage = Stage.activeStages.First();
+
+			characters = new Dictionary<string, Character>();
+		
+			Character[] characterObjects = GameObject.FindObjectsOfType<Fungus.Character>();
+			foreach (Character character in characterObjects)
+			{
+				characters.Add(character.name, character);
+				characters.Add(character.nameText, character);
+			}
+
+			characterNotFound = new Character();
+			characterNotFound.nameText = "Character Not Found";
+
+			characterNone = new Character();
+		}
+
+		/// <summary>
+		/// Parse and execute a conversation string
+		/// </summary>
+		/// <param name="conv"></param>
+		public void Conversation(string conv)
+		{
+			if (string.IsNullOrEmpty(conv)) return;
+			
+			SayDialog sayDialog = SayDialog.GetSayDialog();
+
+			if (sayDialog == null) return;
+
+			//find SimpleScript say strings with portrait options
+			//You can test regex matches here: http://regexstorm.net/tester
+			var sayRegex = new Regex(@"(?<character>\w*)( ?(?<mood>\w*)( ?(?<location>\w*)))?:(?<text>.*)");
+			var sayMatches = sayRegex.Matches(conv);
+
+			foreach (Match match in sayMatches)
+			{
+				string characterKey = match.Groups["character"].Value;
+				string mood = match.Groups["mood"].Value;
+				string location = match.Groups["location"].Value;
+				string text = match.Groups["text"].Value;
+
+				UnityEngine.Debug.Log(text);
+				sayDialog.gameObject.SetActive(true);
+
+				Character character = CharacterSearch(characterKey);
+				sayDialog.SetCharacter(character);
+
+				sayDialog.Say(text, true, true, true, false, null, () => { });
+			}
+		}
+
+		public Character CharacterSearch(string search)
+		{
+			Character character = characters.Where(
+				item => item.Key.StartsWith(search, true, System.Globalization.CultureInfo.CurrentCulture))
+				.FirstOrDefault().Value;
+			if (character == null)
```

Comment:
I'm trying to avoid using Linq these days as it performs loads of memory allocations which can cause gc issues. I should have been tighter on this stuff from the start myself, but I'm gradually improving systems to avoid unneccessary allocations at runtime (e.g. the recent StringBuilder changes to Writer.cs). No problem leaving it using Linq until you get everything working, but lets do an optimisation pass at the end to tidy up. btw allocations at scene start are not as big a deal, it's the ones at runtime we need to worry about!